File: si618w16_project_vedula.py
import pandas as pd 
from dateutil import parser
import csv
import re

def jfk_time(pickup_day=[],pickup_hour=[]):
    for i in range(len(jfk.index.tolist())):
        pickup_day.append(parser.parse(jfk.iloc[i,0]).strftime('%A'))
        pickup_hour.append(str(re.search(r'\d+:',jfk.iloc[i,0]).group()).rstrip(':'))

    return(pickup_day,pickup_hour)

def newark_time(pickup_day=[],pickup_hour=[]):
    for i in range(len(newark.index.tolist())):
        pickup_day.append(parser.parse(newark.iloc[i,0]).strftime('%A'))
        pickup_hour.append(str(re.search(r'\d+:',newark.iloc[i,0]).group()).rstrip(':'))

    return(pickup_day,pickup_hour)

def jfk_output(pickup_day,pickup_hour,pickup_datetime,dropoff_datetime):
    time_delta=[]
    for i in range(len(pickup_datetime)):
        start=parser.parse(pickup_datetime.iloc[i])
        end=parser.parse(dropoff_datetime.iloc[i])
        time_delta.append(((end-start).total_seconds())/60.0)
    data1={'pickup_day':pickup_day,'time_delta':time_delta}
    temp1=pd.DataFrame(data=data1)
    data2={'pickup_hour':pickup_hour,'time_delta':time_delta}
    temp2=pd.DataFrame(data=data2)
    
    p={}
    q={}

    grouped_day=temp1.groupby('pickup_day')
    for day, group in grouped_day:
        p[day]=group['time_delta'].mean()
    
    grouped_hour=temp2.groupby('pickup_hour')
    for hour, group in grouped_hour:
        q[hour]=group['time_delta'].mean()
    
    pd.DataFrame(p.items()).to_csv('mean_time_day_jfk.csv',header=['pickup_day','mean_time'])
    pd.DataFrame(q.items()).to_csv('mean_time_hour_jfk.csv',header=['pickup_hour','mean_time'])

    return 0

def newark_output(pickup_day,pickup_hour,pickup_datetime,dropoff_datetime):
    time_delta=[]
    for i in range(len(pickup_datetime)):
        start=parser.parse(pickup_datetime.iloc[i])
        end=parser.parse(dropoff_datetime.iloc[i])
        time_delta.append(((end-start).total_seconds())/60.0)
    data1={'pickup_day':pickup_day,'time_delta':time_delta}
    temp1=pd.DataFrame(data=data1)
    data2={'pickup_hour':pickup_hour,'time_delta':time_delta}
    temp2=pd.DataFrame(data=data2)
    
    p={}
    q={}

    grouped_day=temp1.groupby('pickup_day')
    for day, group in grouped_day:
        p[day]=group['time_delta'].mean()
    
    grouped_hour=temp2.groupby('pickup_hour')
    for hour, group in grouped_hour:
        q[hour]=group['time_delta'].mean()
    
    pd.DataFrame(p.items()).to_csv('mean_time_day_newark.csv',header=['pickup_day','mean_time'])
    pd.DataFrame(q.items()).to_csv('mean_time_hour_newark.csv',header=['pickup_hour','mean_time'])

    return 0

if __name__ == '__main__':
    trip_data=pd.read_csv('yellow_tripdata.csv')
    trip_data.dropna(axis=0,how='any')
    jfk=trip_data[(trip_data['pickup_latitude']<=40.771 ) & \
             (trip_data['pickup_latitude']>=40.753) & \
             (trip_data['RateCodeID']==2)][[1,2]]
    
    print 'jfk read'
   
    ############################################
    newark=trip_data[(trip_data['pickup_latitude']<=40.771 ) & \
             (trip_data['pickup_latitude']>=40.753) & \
             (trip_data['RateCodeID']==3)][[1,2]]
   
    print 'Newark read'
    #############################################
    jfk_pickup_day,jfk_pickup_hour=jfk_time()

    print "Jfk time processed"
    #############################################
    newark_pickup_day,newark_pickup_hour=newark_time()

    print 'Newark time processed'
    #############################################
    jfk_output(jfk_pickup_day,jfk_pickup_hour,jfk['tpep_pickup_datetime'],jfk['tpep_dropoff_datetime'])
    
    print 'jfk output printed'
    #############################################
    newark_output(newark_pickup_day,newark_pickup_hour,newark['tpep_pickup_datetime'],newark['tpep_dropoff_datetime'])
    
    print 'Newark output printed'
    #############################################
    late_night=trip_data[['pickup_latitude','pickup_longitude','tpep_pickup_datetime']]
    #print late_night.head()
    count=0
    with open('latenight_tripdata.csv','w') as csvfile:
        fieldnames=['pickup_latitude','pickup_longitude','pickup_hour']
        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
        writer.writeheader()
        for i in range(len(late_night)):
            if ((parser.parse(late_night.iloc[i,2]).hour>=22) or (parser.parse(late_night.iloc[i,2]).hour<5)):
                writer.writerow({'pickup_latitude': late_night.iloc[i]['pickup_latitude'],\
                               'pickup_longitude': late_night.iloc[i]['pickup_longitude'],\
                               'pickup_hour':str(re.search(r'\d+:',late_night.iloc[i]['tpep_pickup_datetime']).group()).rstrip(':')})
            count+=1
            if count>350000:
                break

    print 'late night trip data printed'
    #############################################
    pd.DataFrame(trip_data[['payment_type','fare_amount','tip_amount']]).to_csv('tip.csv')
    
    print 'tip amount printed'
    #############################################
    data=trip_data[(trip_data['dropoff_latitude']>=40.714886) & \
                  (trip_data['dropoff_latitude']<=40.714916)][['tpep_dropoff_datetime']]
    for i in range(len(data)):
        data.iloc[i,0]= str(re.search(r'\d+:',data.iloc[i,0]).group()).rstrip(':')
    grouped=data.groupby('tpep_dropoff_datetime')
    p={}
    for hour,group in grouped:
        p[hour]=len(group)
    pd.DataFrame(p.items()).to_csv('goldman_sachs.csv',header=['dropoff_hour','total_dropoffs'])

    print 'goldman sachs printed'

File: si618w16_project_vedula.R
library('ggplot2')
library('ggmap')
jfk_day=read.csv('mean_time_day_jfk.csv',header=T)
jfk_day$pickup_day=factor(jfk_day$pickup_day,
  levels=c('Sunday','Monday','Tuesday','Wednesday','Thursday','Friday','Saturday'))
jfk_day=jfk_day[order(jfk_day$pickup_day),]

graph1=ggplot(jfk_day,aes(pickup_day,mean_time))+
  geom_bar(stat='identity',width=0.9)+
  ggtitle('Average time to reach JFK international airport')+
  labs(x='Day',y='Average time taken in minutes')+
  theme(axis.text.x=element_text(angle=60,vjust=0.5))
print (graph1)
#####################################################
jfk_hour=read.csv('mean_time_hour_jfk.csv',header=T)
jfk_hour=jfk_hour[order(jfk_hour$pickup_hour),]

graph2=ggplot(jfk_hour,aes(pickup_hour,mean_time))+
  geom_line()+
  ggtitle('Average time to reach JFK international airport')+
  labs(x='Pickup hour in 24 hour format',y='Average time taken in minutes')+
  theme(axis.text.x=element_text(vjust=0.5))
print (graph2)
#####################################################
newark_day=read.csv('mean_time_day_newark.csv',header=T)
newark_day$pickup_day=factor(newark_day$pickup_day,
                          levels=c('Sunday','Monday','Tuesday','Wednesday','Thursday','Friday','Saturday'))
newark_day=newark_day[order(newark_day$pickup_day),]

graph3=ggplot(newark_day,aes(pickup_day,mean_time))+
  geom_bar(stat='identity',width=0.9)+
  ggtitle('Average time to reach Newark airport')+
  labs(x='Day',y='Average time taken in minutes')+
  theme(axis.text.x=element_text(angle=60,vjust=0.5))
print (graph3)
###########################################################
newark_hour=read.csv('mean_time_hour_newark.csv',header=T)
newark_data1=newark_hour[newark_hour$pickup_hour==22,]

graph4=ggplot(newark_hour,aes(pickup_hour,mean_time))+
  geom_line(stat='identity')+
  ggtitle('Average time to reach Newark airport')+
  labs(x='Pickup hour in 24 hour format',y='Average time taken in minutes')+
  theme(axis.text.x=element_text(vjust=0.5))
print (graph4)
###########################################################
latenight_data=read.csv('latenight_tripdata.csv',header=T)
latenight_data=latenight_data[order(latenight_data$pickup_hour),]

nyc=get_map(location = 'New York City', zoom = 13,source='google')
ggmap(nyc)+geom_density2d(data =latenight_data, aes(x=pickup_longitude, 
      y = pickup_latitude),size=0.3) +
      stat_density2d(data = latenight_data, 
      aes(x =pickup_longitude, y = pickup_latitude, 
      fill = ..level.., alpha = ..level..), size = 0.05,bins = 20, geom = "polygon")+
      scale_fill_gradient(low = "yellow", high = "red")+
      scale_alpha(range = c(0, 0.5), guide = FALSE)+
      labs(y ='Latitude', x ='Longitude')+
      ggtitle('Pickup Activity between 10 PM - 5 AM ')
##########################################################
tips=read.csv('tip.csv',header=T)
tips=subset(tips,(payment_type==1 | payment_type==2))
data1=subset(tips,fare_amount>=1 & fare_amount<10)

graph6=ggplot(data1,aes(fare_amount,tip_amount))+
       stat_binhex()+geom_smooth(method=lm,se=F,color='green')+
       ggtitle('Tpping Trend')+labs(x='Fare amount in Dollars', y='Tip amount')+
       theme(axis.text.x=element_text(vjust=0.5))+facet_wrap(~payment_type)
print (graph6)

data2=subset(tips,fare_amount>=10 & fare_amount<20)

graph61=ggplot(data2,aes(fare_amount,tip_amount))+
  stat_binhex()+geom_smooth(method=lm,se=F,color='green')+
  ggtitle('Tpping Trend')+labs(x='Fare amount in Dollars', y='Tip amount')+
  theme(axis.text.x=element_text(vjust=0.5))+facet_wrap(~payment_type)
print (graph61)

data3=subset(tips,fare_amount>=20 & tip_amount<600)

graph62=ggplot(data3,aes(fare_amount,tip_amount))+
  stat_binhex()+geom_smooth(method=lm,se=F,color='green')+
  ggtitle('Tpping Trend')+labs(x='Fare amount in Dollars', y='Tip amount')+
  theme(axis.text.x=element_text(vjust=0.5))+facet_wrap(~payment_type)
print (graph62)
##########################################################
goldman=read.csv('goldman_sachs.csv',header=T)

graph7=ggplot(goldman,aes(dropoff_hour,total_dropoffs))+
  geom_bar(stat='identity',width=0.95)+
  ggtitle('Total number of dropoffs at Goldman Sachs')+
  labs(x='Dropoff hour',y='Total number of dropoffs')+
  theme(axis.text.x=element_text(angle=60,vjust=0.5))
print (graph7)
